pkg/ottl/functions.go

```diff
@@ -70,13 +97,60 @@ func (p *Parser[K]) buildArgs(ed editor, argsVal reflect.Value) error {
 			return fmt.Errorf("ottlarg struct tag on field %q is not a valid integer: %w", argsType.Field(i).Name, err)
 		}
 
-		if argNum < 0 || argNum >= len(ed.Arguments) {
-			return fmt.Errorf("ottlarg struct tag on field %q has value %d, but must be between 0 and %d", argsType.Field(i).Name, argNum, len(ed.Arguments))
+		if argNum < 0 || argNum >= len(ed.NewArguments) {
+			return fmt.Errorf("ottlarg struct tag on field %q has value %d, but must be between 0 and %d", argsType.Field(i).Name, argNum, len(ed.NewArguments))
+		}
+
+		argVal := ed.NewArguments[argNum].Value
+
+		var val any
+		var fieldVal reflect.Value
+		if isOptional {
+			fieldVal = field.FieldByName("Val")
+			fieldType = fieldVal.Type()
+			isOptional = true
+		} else {
+			requiredArgs++
+		}
+
+		if fieldType.Kind() == reflect.Slice {
+			val, err = p.buildSliceArg(argVal, fieldType)
+		} else {
+			val, err = p.buildArg(argVal, fieldType)
+		}
+
+		if err != nil {
+			return fmt.Errorf("invalid argument at position %v: %w", i, err)
 		}
+		if isOptional {
+			fieldVal.Set(reflect.ValueOf(val))
```

Comment:
what happens when an optional param is not passed in, would this be `nil`?

Reply:
The `ottl.Optional` type will indicate (currently through `HasValue` or `IsEmpty`) whether a value is present, but the field value here will be the zero value of the underlying type, which usually won't be `nil`. The existence of zero values in Go necessitates having another way to indicate whether a value was passed in. We could potentiall use pointers here (which have `nil` as a zero value), but I would prefer to have something that is more explicit.